vercel/turborepo
# Added "outputLogs" Configuration Option

### Description

This pull request builds on #822 by adding an `"outputLogs"` option to `turbo.json`. We've found that the default behavior is very noisy and doesn't provide any value to us. Rather than appending `--output-logs` every time we use a command, it would be nice to set the configuration to do this by default.

* Defaults to "full" in order to maintain backward compatibility with the current behavior.
* `--output-logs` overrides the configuration option.
* Merged `cacheHitLogsMode` and `cacheMissLogsMode` into a single type. Given the logic around this behavior, it seemed unnecessary to maintain two separate internal options when they provided no clear benefit.
* Refactored `LogsMode` into the `util` package to avoid circular dependencies in `run`, `runcache`, and `fs`.
  * I went with the `TaskOutputMode` naming because it feels like we're talking more about the task output than the log output. As far as the end-user is concerned, whether it's cached or not, this is the output of the `package.json` task.
* Added `"outputLogs": "full"` to root `turbo.json` for clarity.
* Documented the new configuration option and override behavior for `--output-logs`.

### Questions

* Should we name this `"taskOutputMode"` for consistency and a better description of what it does?
* Should this be a task configuration rather than a top-level configuration? It would be a trivial change to make in this PR and would support having a configuration where a `"build"` task has no output but another does. I'm not sure if there's any use to that though?

### Testing Instructions

1. Check `./turbow.sh run build`, behavior should match the current
2. Change `"outputLogs"` in `turbo.json` to `"none"`
3. Check `./turbow.sh run build`, behavior should match current with `--output-logs=none`
4. Change `"outputLogs"` in `turbo.json` to `"hash-only"`
5. Check `./turbow.sh run build`, behavior should match current with `--output-logs=hash-only`
4. Change `"outputLogs"` in `turbo.json` to `"new-only"`
5. Check `./turbow.sh run build`, behavior should match current with `--output-logs=new-only`
6. Check `./turbow.sh run build --output-logs=none`, behavior should match current with `--output-logs=none`
7. Optionally check other `--output-logs` inputs and verify that they override the configuration and match the current
